cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,282 @@
+    }
+  }
+
+  private class ConstraintCheckerThread extends Thread {
+    private final RetryStrategy scheduleStrategy;
+    private int failureCount;
+
+
+    ConstraintCheckerThread() {
+      super(String.format("ConstraintCheckerService-%s", "id"));
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+    }
+
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = checkJobQueue();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Check jobs in job queue for constraint satisfaction.
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long checkJobQueue() {
+      boolean emptyFetch = false;
+      try {
+        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            return checkJobConstraints(context);
+          }
+        });
+        failureCount = 0;
+      } catch (Exception e) {
+        LOG.warn("Failed to check Job constraints. Will retry in next run", e);
+        failureCount++;
+      }
+
+      try {
+        Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            // run any ready jobs
+            return runReadyJobs(context);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to launch programs. Will retry in next run", e);
+        failureCount++;
+      }
+
+      // If there is any failure, delay the next fetch based on the strategy
+      if (failureCount > 0) {
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+
+      // Sleep for 2 seconds if there's no jobs in the queue
+      return emptyFetch ? 2000L : 0L;
+    }
+
+    private boolean checkJobConstraints(DatasetContext context) throws Exception {
+      boolean emptyScan = true;
+
+      JobQueueDataset jobQueue = getJobQueue(context);
+      try (CloseableIterator<Job> jobsIter = jobQueue.getAllJobs()) {
+        while (jobsIter.hasNext() && isRunning()) {
+          emptyScan = false;
+          Job job = jobsIter.next();
+          switch (job.getState()) {
+            case PENDING_TRIGGER:
+              if (isTriggerSatisfied(job)) {
+                job.setState(Job.State.PENDING_CONSTRAINT);
```

Comment:
would it be better to have a separate column for the job state? So that you don't need to update entire Json (whole job including the schedule and notifications)?